src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,229 @@
+by { rw [←circulant_single], congr, ext i, simp [pi.single, update] }
+
+/-- Another version of `one_eq_circulant`. -/
+lemma one_eq_circulant' [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, (1 : matrix I I α) i 0) :=
+by { ext i j, simp only [circulant, one_apply, sub_eq_zero], congr }
+
+lemma fin.one_eq_circulant [has_zero α] [has_one α] :
+  ∀ n, (1 : matrix (fin n) (fin n) α) = circulant (λ i, ite (↑i = 0) 1 0)
+| 0     := dec_trivial
+| (n+1) :=
+begin
+  ext i j,
+  simp only [circulant, one_apply, ← @fin.coe_zero n, fin.coe_injective.eq_iff, sub_eq_zero],
+  congr,
+end
+
+/-- For a one-ary predicate `p`, `p` applied to every entry of `circulant v` is true
+    if `p` applied to every entry of `v` is true. -/
+lemma pred_circulant_entry_of_pred_vec_entry [has_sub I]
+  {p : α → Prop} {v : I → α} (h : ∀ k, p (v k)) (i j : I) :
+  p ((circulant v) i j) := h (i - j)
+
+/-- Given a set `S`, every entry of `circulant v` is in `S` if every entry of `v` is in `S`. -/
+lemma circulant_entry_in_of_vec_entry_in [has_sub I]
+  {S : set α} {v : I → α} (h : ∀ k, v k ∈ S) (i j : I) :
+  (circulant v) i j ∈ S := h (i - j)
+
+/-- The circulant matrix `circulant v` is symmetric iff `∀ i j, v (j - i) = v (i - j)`. -/
+lemma circulant_is_symm_ext_iff' [has_sub I] {v : I → α} :
+  (circulant v).is_symm ↔ ∀ i j, v (j - i) = v (i - j) :=
+by simp [is_symm.ext_iff, circulant]
+
+/-- The circulant matrix `circulant v` is symmetric iff `v (- i) = v i` if `[add_group I]`. -/
+lemma circulant_is_symm_ext_iff [add_group I] {v : I → α} :
+  (circulant v).is_symm ↔ ∀ i, v (- i) = v i :=
+begin
+  rw [circulant_is_symm_ext_iff'],
+  split,
+  { intros h i, convert h i 0; simp },
+  { intros h i j, convert h (i - j), simp }
+end
+
+lemma fin.circulant_is_symm_ext_iff :
+  ∀ {n} {v : fin n → α}, (circulant v).is_symm ↔ ∀ i, v (- i) = v i
+| 0     := λ v, by simp only [circulant_is_symm_ext_iff', is_empty.forall_iff]
+| (n+1) := λ v, circulant_is_symm_ext_iff
+
+/-- If `circulant v` is symmetric, `∀ i j : I, v (j - i) = v (i - j)`. -/
+lemma circulant_is_symm_apply' [has_sub I] {v : I → α} (h : (circulant v).is_symm) (i j : I) :
+  v (j - i) = v (i - j) :=
+circulant_is_symm_ext_iff'.1 h i j
+
+/-- If `circulant v` is symmetric, `∀ i j : I, v (- i) = v i`. -/
+lemma circulant_is_symm_apply [add_group I] {v : I → α} (h : (circulant v).is_symm) (i : I) :
+  v (-i) = v i :=
+circulant_is_symm_ext_iff.1 h i
+
+lemma fin.circulant_is_symm_apply {v : fin n → α} (h : (circulant v).is_symm) (i : fin n) :
+  v (-i) = v i :=
+fin.circulant_is_symm_ext_iff.1 h i
+
+/-- The associated polynomial `(v 0) + (v 1) * X + ... + (v (n-1)) * X ^ (n-1)` to `circulant v`.-/
+noncomputable
+def circulant.poly [semiring α] (v : fin n → α) : polynomial α :=
+∑ i : fin n, polynomial.monomial i (v i)
+
+/-- `circulant_perm n` is the cyclic permutation over `fin n`. -/
+def circulant.perm : Π n, equiv.perm (fin n) := λ n, equiv.symm (fin_rotate n)
```

Comment:
Again, arguments are usually better before the colon:
```suggestion
def circulant.perm (n : ℕ) : equiv.perm (fin n) := (fin_rotate n).symm
```
This of course suggests that there is no point having this definition in the first place.